File: modules/keypairs/keypairs.go
// MITHRAS: Javascript configuration management tool for AWS.
// Copyright (C) 2016, Colin Steele
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//                  (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//     but WITHOUT ANY WARRANTY; without even the implied warranty of
//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//              GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
package keypairs

// @public
//
// This package exports several entry points into the JS environment,
// including:
//
// > * [aws.keypairs.scan](#vscan)
// > * [aws.keypairs.create](#vcreate)
// > * [aws.keypairs.delete](#vdelete)
// > * [aws.keypairs.describe](#vdescribe)
//
// This API allows resource handlers to manage AWS SSH keypairs.
//
// ## AWS.KEYPAIRS.SCAN
// <a name="vscan"></a>
// `aws.keypairs.scan(region);`
//
// Returns a list of keypairs.
//
// Example:
//
// ```
//
//  var keypairs =  aws.keypairs.scan("us-east-1");
//
// ```
//
// ## AWS.KEYPAIRS.CREATE
// <a name="vcreate"></a>
// `aws.keypairs.create(region, config);`
//
// Create a keypair.
//
// Example:
//
// ```
//
//  var keypair =  aws.keypairs.create(
//    "us-east-1",
//    {
//		  KeyName: "my-keypair"
//    });
//
// ```
//
// ## AWS.KEYPAIRS.DELETE
// <a name="vdelete"></a>
// `aws.keypairs.delete(region, keypair_id);`
//
// Delete a keypair.
//
// Example:
//
// ```
//
//  aws.keypairs.delete("us-east-1", "my-keypair");
//
// ```
//
// ## AWS.KEYPAIRS.DESCRIBE
// <a name="vdescribe"></a>
// `aws.keypairs.describe(region, keypair_id);`
//
// Get info from AWS about a KEYPAIR.
//
// Example:
//
// ```
//
//  var keypair = aws.keypairs.describe("us-east-1", "my-keypair");
//
// ```
//

import (
	log "github.com/Sirupsen/logrus"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/ec2"
	"github.com/robertkrimen/otto"

	mcore "github.com/cvillecsteele/mithras/modules/core"
)

var Version = "1.0.0"
var ModuleName = "keypairs"

func describe(region string, id string) *ec2.KeyPairInfo {
	svc := ec2.New(session.New(),
		aws.NewConfig().WithRegion(region).WithMaxRetries(5))

	params := &ec2.DescribeKeyPairsInput{
		KeyNames: []*string{
			aws.String(id),
		},
	}
	resp, err := svc.DescribeKeyPairs(params)

	if err != nil {
		return nil
	}
	if len(resp.KeyPairs) > 0 {
		return resp.KeyPairs[0]
	}
	return nil
}

func create(region string, name string) string {
	svc := ec2.New(session.New(),
		aws.NewConfig().WithRegion(region).WithMaxRetries(5))

	params := &ec2.CreateKeyPairInput{
		KeyName: aws.String(name),
	}
	resp, err := svc.CreateKeyPair(params)
	if err != nil {
		log.Fatalf("Error creating key: %s", err)
	}

	return *resp.KeyMaterial
}

func delete(region string, id string) {
	svc := ec2.New(session.New(),
		aws.NewConfig().WithRegion(region).WithMaxRetries(5))

	params := &ec2.DeleteKeyPairInput{
		KeyName: aws.String(id),
	}
	_, err := svc.DeleteKeyPair(params)

	if err != nil {
		log.Fatalf("Error deleting key: %s", err)
	}
}

func scan(region string) []*ec2.KeyPairInfo {
	svc := ec2.New(session.New(),
		aws.NewConfig().WithRegion(region).WithMaxRetries(5))

	params := &ec2.DescribeKeyPairsInput{}
	resp, err := svc.DescribeKeyPairs(params)

	if err != nil {
		log.Fatalf("Error scanning keys: %s", err)
	}

	return resp.KeyPairs
}

func init() {
	mcore.RegisterInit(func(context *mcore.Context) {
		rt := context.Runtime

		if a, err := rt.Get("aws"); err != nil || a.IsUndefined() {
			rt.Object(`aws = {}`)
		}
		o1, _ := rt.Object(`aws.keypairs = {}`)
		o1.Set("scan", func(region string) otto.Value {
			f := mcore.Sanitizer(rt)
			return f(scan(region))
		})
		o1.Set("create", func(region, id string) otto.Value {
			f := mcore.Sanitizer(rt)
			return f(create(region, id))
		})
		o1.Set("delete", func(region, id string) otto.Value {
			delete(region, id)
			return otto.Value{}
		})
		o1.Set("describe", func(region, id string) otto.Value {
			f := mcore.Sanitizer(rt)
			return f(describe(region, id))
		})
	})
}

File: js/keypairs.js
// MITHRAS: Javascript configuration management tool for AWS.
// Copyright (C) 2016, Colin Steele
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//                  (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//     but WITHOUT ANY WARRANTY; without even the implied warranty of
//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//              GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// @public
// 
// # keypairs
// 
// Keypairs is resource handler for manipulating AWS SSH keypairs.
// 
// This module exports:
// 
// > * `init` Initialization function, registers itself as a resource
// >   handler with `mithras.modules.handlers` for resources with a
// >   module value of `"keypairs"`
// 
// Usage:
// 
// `var keypairs = require("keypairs").init();`
// 
//  ## Example Resource
// 
// ```javascript
// var rKey = {
//      name: "key"
//      module: "keypairs"
//      skip: (ensure === 'absent') // Don't delete keys
//      params: {
//          region: "us-east-1"
//          ensure: "present"
//          key: {
//              KeyName: "my-fancy-key"
//          }
//          savePath: os.expandEnv("$HOME/.ssh/" + keyName + ".pem")
//      }
// };
// ```
// 
// ## Parameter Properties
// 
// ### `ensure`
//
// * Required: true
// * Allowed Values: "absent", "present"
//
// If `"present"`, the handler will ensure the keypair exists, and it
// not, it will be created.  If `"absent"`, the keypair is removed.
// 
// ### `key`
//
// * Required: true
// * Allowed Values: JSON corresponding to the structure found [here](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2.html#type-CreateKeyPairInput)
//
// Specifies parameters for keypair creation.
//
// ### `savepath`
//
// * Required: true
// * Allowed Values: A valid path for saving the pemfile when a keypair is created
//
// When the handler creates a new keypair, the contents of the key are saved to this path.
//
// ### `on_find`
//
// * Required: true
// * Allowed Values: A function taking two parameters: `catalog` and `resource`
//
// If defined in the resource's `params` object, the `on_find`
// function provides a way for a matching resource to be identified
// using a user-defined way.  The function is called with the current
// `catalog`, as well as the `resource` object itself.  The function
// can look through the catalog, find a matching object using whatever
// logic you want, and return it.  If the function returns `undefined`
// or a n empty Javascript array, (`[]`), the function is indicating
// that no matching resource was found in the `catalog`.
// 
(function (root, factory){
    if (typeof module === 'object' && typeof module.exports === 'object') {
        module.exports = factory();
    }
})(this, function() {
    
    var sprintf = require("sprintf.js").sprintf;

    var handler = {
        moduleNames: ["keypairs"]
        findInCatalog: function(catalog, resource) {
            if (typeof(resource.params.on_find) === 'function') {
		result = resource.params.on_find(catalog, resource);
		if (!result || 
		    (Array.isArray(result) && result.length == 0)) {
		    return;
		}
		return result;
	    }
            return _.find(catalog.keypairs, function(key) { 
                return key.KeyName === resource.params.key.KeyName;
            });
        }
        handle: function(catalog, resources, resource) {
            if (!_.find(handler.moduleNames, function(m) { return resource.module === m; })) {
                return [null, false];
            }
                
            var ensure = resource.params.ensure;
            var params = resource.params;
            
            if (!params.ensure || !params.key || !params.savePath || !params.region) {
                console.log("Invalid key params")
                os.exit(3);
            }
            var found = resource._target;

            switch(ensure) {
            case "absent":
                if (!found) {
                    if (mithras.verbose) {
                        log(sprintf("Key '%s' not found, no action taken.", params.key.KeyName));
                        break;
                    }
                }
                if (mithras.verbose) {
                    log(sprintf("Deleting key '%s'", params.key.KeyName));
                }
                aws.keypairs.delete(params.region, params.key.KeyName);
                catalog.keypairs = 
                    _.reject(catalog.keypairs,
                             function(k) { 
                                 return k.KeyName == params.key.KeyName;
                             });
                break;
            case "present":
                if (found) {
                    if (mithras.verbose) {
                        log(sprintf("Key '%s' found, no action taken.", params.key.KeyName));
                        break;
                    }
                }

                if (mithras.verbose) {
                    log(sprintf("Creating key '%s'", params.key.KeyName));
                }
                var raw = aws.keypairs.create(params.region, params.key.KeyName);
                if (mithras.verbose) {
                    log(sprintf("Writing key '%s' to '%s'", params.key.KeyName, params.savePath));
                }
                var err = fs.write(params.savePath, raw, 0400);
		if (err) {
                    log(sprintf("Error writing key '%s' to '%s': %s", 
				params.key.KeyName, 
				params.savePath,
				err));
		    os.exit(3);
		}

                var key = aws.keypairs.describe(params.region, params.key.KeyName);
                do {
                    time.sleep(10);
                    key = aws.keypairs.describe(params.region, params.key.KeyName);
                } while (!key);
                catalog.keypairs.push(key);

                // return it
                return [handler.findInCatalog(catalog, resource), true];
                break;
            }
            return [null, true];
        }
        preflight: function(catalog, resources, resource) {
            if (!_.find(handler.moduleNames, function(m) { 
                return resource.module === m; 
            })) {
                return [null, false];
            }
            var found = handler.findInCatalog(catalog, resource);
            if (found) {
                return [found, true];
            }
            return [null, true];
        }
    };
    
    handler.init = function () {
        mithras.modules.preflight.register(handler.moduleNames[0], handler.preflight);
        mithras.modules.handlers.register(handler.moduleNames[0], handler.handle);
        return handler;
    };
    
    return handler;
});
